compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3232,7 +3232,8 @@ private void checkFieldBasedAccess(BLangFieldBasedAccess fieldAccessExpr, boolea
             if (actualType != symTable.semanticError &&
                     (fieldAccessExpr.isLValue || fieldAccessExpr.isCompoundAssignmentLValue)) {
                 if (isAllReadonlyTypes(varRefType)) {
-                    if (varRefType.tag != TypeTags.OBJECT || !isInitializationInInit(varRefType, data)) {
+                    if (Types.getReferredType(varRefType).tag != TypeTags.OBJECT
+                            || !isInitializationInInit(varRefType, data)) {
```

Comment:
`isInitializationInInit` does a `BObjectType` cast. Hence, need the referred type there